src/frontend/flight/sync.rs

```diff
@@ -376,3 +432,109 @@ fn now() -> u64 {
         .expect("Time went backwards")
         .as_secs()
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::context::test_utils::in_memory_context;
+    use crate::frontend::flight::sync::SeafowlDataSyncManager;
+    use crate::frontend::flight::SEAFOWL_SYNC_DATA_UD_FLAG;
+    use arrow::array::{BooleanArray, Int32Array, RecordBatch, StringArray};
+    use arrow_schema::{DataType, Field, Schema};
+    use std::sync::Arc;
+    use std::time::Duration;
+    use uuid::Uuid;
+
+    fn dummy_batch() -> RecordBatch {
+        let schema = Arc::new(Schema::new(vec![
+            Field::new("c1", DataType::Int32, true),
+            Field::new("c2", DataType::Utf8, true),
+            Field::new(SEAFOWL_SYNC_DATA_UD_FLAG, DataType::Boolean, false),
+        ]));
+
+        RecordBatch::try_new(
+            schema.clone(),
+            vec![
+                Arc::new(Int32Array::from(vec![1, 2, 3])),
+                Arc::new(StringArray::from(vec![Some("one"), None, Some("three")])),
+                Arc::new(BooleanArray::from(vec![true, true, true])),
+            ],
+        )
+        .unwrap()
+    }
+
+    #[tokio::test]
+    async fn test_flush_interleaving() {
+        let mut ctx = in_memory_context().await;
+
+        // Just enough for two dummy batches, so a flush is forced after the third
+        ctx.config.misc.sync_data.max_in_memory_bytes = 3000;
+        // Force the remaining flush after sleeping for 1 second
+        ctx.config.misc.sync_data.max_replication_lag_s = 1;
+
+        let table_1_uuid = Uuid::new_v4();
+        let table_2_uuid = Uuid::new_v4();
+
+        let log_store_1 = ctx
+            .internal_object_store
+            .get_log_store(&table_1_uuid.to_string());
+        let log_store_2 = ctx
+            .internal_object_store
+            .get_log_store(&table_2_uuid.to_string());
+
+        let mut sync_mgr = SeafowlDataSyncManager::new(Arc::new(ctx));
+
+        let origin = "origin".to_string();
+
+        // Add LSN 1 for table 1
+        sync_mgr
+            .sync_batches(
+                log_store_1.clone(),
+                1,
+                "origin".to_string(),
+                vec!["c1".to_string()],
+                true,
+                vec![dummy_batch()],
+            )
+            .await
+            .unwrap();
+
+        assert_eq!(sync_mgr.stored_sequences(&origin), (Some(1), None),);
+
+        // Add LSN 2 for table 2
+        sync_mgr
+            .sync_batches(
+                log_store_2.clone(),
+                2,
+                "origin".to_string(),
+                vec!["c1".to_string()],
+                true,
+                vec![dummy_batch()],
+            )
+            .await
+            .unwrap();
+
+        assert_eq!(sync_mgr.stored_sequences(&origin), (Some(2), None),);
+
+        // Add LSN 3 for table 1 and push it over the max configured size
```

Comment:
What would happen if you added a change for table_2 instead (this would also trigger a flush) - would we flush just the changes for table_1 and advance the LSN to 1?